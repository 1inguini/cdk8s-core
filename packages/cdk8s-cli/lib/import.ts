import * as https from 'https';
import { promises } from 'fs';
import { emitConstructForApiObject, findApiObjectDefinitions } from '../lib/codegen-constructs';
import { CodeMaker } from 'codemaker';
import { JSONSchema4 } from 'json-schema';

export interface Options {
  readonly apiVersion: string;
}

export async function generateAllApiObjects(outdir: string, options: Options) {
  const code = new CodeMaker();
  code.indentation = 2;

  const schema = await downloadSchema(options.apiVersion);
  const topLevelObjects = findApiObjectDefinitions(schema);

  for (const o of topLevelObjects) {
    emitConstructForApiObject(code, schema, o);
  }

  await promises.mkdir(outdir, { recursive: true });
  await code.save(outdir);  
}

async function downloadSchema(apiVersion: string) {
  const output = await httpsGet(`https://kubernetesjsonschema.dev/v${apiVersion}/_definitions.json`);
  return JSON.parse(output) as JSONSchema4;
}

async function httpsGet(url: string): Promise<string> {
  return new Promise((ok, ko) => {
    const req = https.get(url, res => {
      const data = new Array<Buffer>();
      res.on('data', chunk => data.push(chunk));
      res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
      res.once('error', ko);
    });

    req.once('error', ko);
    req.end();
  });
}
