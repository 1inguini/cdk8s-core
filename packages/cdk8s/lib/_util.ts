export function removeEmpty(obj: any): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }

  if (Object.keys(obj).length === 0) {
    return undefined;
  }
  
  if (Array.isArray(obj)) {
    if (obj.length === 0) {
      return undefined;
    }

    return obj.map(x => {
      // do not remove "{}" and "[]" if they are array elements.
      if (x != null && typeof(x) === 'object' && Object.keys(x).length === 0) { return x; }
      if (Array.isArray(x) && x.length === 0) { return x; }
      return removeEmpty(x)
    });
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple objects: ${obj.constructor.name}`);
  }

  const newObj: { [key: string]: any } = { };

  for (const [key, value] of Object.entries(obj)) {
    const newValue = removeEmpty(value);;
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  return newObj;
};