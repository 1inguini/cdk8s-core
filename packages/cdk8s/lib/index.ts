export * from './service';
export * from './deployment';
export * from './api-object';
export * from './chart';
export * from './horizontal-pod-autoscaler';
export * from './ingress';