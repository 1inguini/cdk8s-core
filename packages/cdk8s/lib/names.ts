import * as crypto from 'crypto';

const MAX_LEN = 253;
const VALIDATE = /^[0-9a-z\-\.]+$/;
const HASH_LEN = 8;

/**
 * Generates a name from a construct path.
 *
 * Kubernetes resources can have names up to 253 characters long. The characters
 * allowed in names are: digits (0-9), lower case letters (a-z), -, and .
 *
 * Names in cdk8s are always in the follow form:
 *
 *  <comp0>.<comp1>....<compN>.<short-hash>
 *
 * Where <comp> are construct node path components.
 *
 * Note that if the total length is longer than 253 characters, we will trim the
 * first components first since the last components usually are more inline with
 * the essence of the resource.
 *
 * @param path the construct path (components separated by "/")
 * @param maxLen maximum allowed length for name
 * @throws if any of the components do not adhere to naming constraints or length.
 */
export function renderObjectName(path: string, maxLen = MAX_LEN) {
  if (maxLen < HASH_LEN) {
    throw new Error(`minimum max length for object names is ${HASH_LEN} (required for hash)`);
  }

  const components = path.split('/');

  // verify components only use allowed chars.
  for (const comp of components) {
    if (!VALIDATE.test(comp)) {
      throw new Error(`"${comp}" is not a valid object name. The characters allowed in names are: digits (0-9), lower case letters (a-z), -, and .`);
    }

    if (comp.length > maxLen) {
      throw new Error(`Object name "${comp}" is too long. Maximum allowed length is ${maxLen}`);
    }
  }

  // special case: if we only have one component in our path, we don't decorate it
  if (components.length === 1) {
    return components[0];
  }

  components.push(calcHash(path, HASH_LEN));

  return components
    .reverse()
    .join('/')
    .slice(0, maxLen)
    .split('/')
    .reverse()
    .join('.');
}

function calcHash(path: string, maxLen: number) {
  const hash = crypto.createHash('sha256');
  hash.update(path);
  return hash.digest('hex').slice(0, maxLen);
}