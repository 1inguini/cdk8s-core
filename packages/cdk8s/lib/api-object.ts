import { Construct } from '@aws-cdk/core';
import { removeEmpty } from './util';
import { renderObjectName } from './names';

export interface ApiObjectMetadata {
  readonly name?: string;
  readonly [key: string]: any;
}


export interface ApiObjectOptions {
/**
   * Data associated with the resource.
   */
  readonly data?: any;

  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ApiObjectMetadata;  

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * Additional attributes for this API object.
   */
  readonly [key: string]: any;
}

export class ApiObject extends Construct {

  /**
   * The app-unique name of the object.
   *
   * The name is allocated based on the path of the object construct within the
   * construct tree.
   */
  public readonly name: string;

  /**
   * The object's API version.
   */
  public readonly apiVersion: string;

  /**
   * The object kind.
   */
  public readonly kind: string;

  constructor(scope: Construct, ns: string, private readonly options: ApiObjectOptions) {
    super(scope, ns);
    this.kind = options.kind;
    this.apiVersion = options.apiVersion;
    this.name = options.metadata?.name ?? this.generateName();
  }

  /**
   * Generates a name for this object based on the construct path. 
   *
   * Kubernetes resources can have names up to 253 characters long. The
   * characters allowed in names are: digits (0-9), lower case letters (a-z), -,
   * and ..
   */
  protected generateName() {
    return renderObjectName(this.node.path);
  }

  /**
   * Renders the object to Kubernetes config.
   */
  public render(): any {
    return removeEmpty({
      ...this.options,
      metadata: {
        ...this.options.metadata,
        name: this.name
      }
    });
  }
}

