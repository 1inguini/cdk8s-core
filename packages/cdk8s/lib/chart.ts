import { Construct, ISynthesisSession } from '@aws-cdk/core';
import fs = require('fs');
import path = require('path');
import { ApiObject } from './api-object';
import YAML = require('yaml');

export class Chart extends Construct {

  /**
   * The name of the stack's YAML file as emitted into the cloud assembly
   * directory during synthesis.
   */
  public readonly manifestFile: string;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.manifestFile = `${this.node.uniqueId}.k8s.yaml`;
  }

  protected synthesize(session: ISynthesisSession) {
    const resources = new Array<any>();

    for (const resource of this.node.findAll()) {
      if (resource instanceof ApiObject) {
        resources.push(resource.render());
      }
    }

    const doc = resources.map(r => toYaml(r)).join('---\n');
    fs.writeFileSync(path.join(session.assembly.outdir, this.manifestFile), doc);
  }
}

function toYaml(o: any) {
  // lose anchors which are based on reference equality
  const x = JSON.parse(JSON.stringify(o));
  return YAML.stringify(x);
}
