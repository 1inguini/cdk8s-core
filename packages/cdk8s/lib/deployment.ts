import { ApiObject } from './api-object';
import { Construct } from '@aws-cdk/core';
import { DeploymentspecAppsV1 } from './.generated/deploymentspec-apps-v1';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';

export interface DeploymentObjectOptions {
  readonly spec: DeploymentspecAppsV1;
  readonly metadata?: ObjectmetaMetaV1;
}

export class DeploymentObject extends ApiObject {
  constructor(scope: Construct, ns: string, options: DeploymentObjectOptions) {
    super(scope, ns, {
      apiVersion: 'apps/v1',
      kind: 'Deployment',
      spec: options.spec,
      metadata: options.metadata
    });
  }
}
