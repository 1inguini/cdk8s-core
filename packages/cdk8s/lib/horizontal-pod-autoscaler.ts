import { ApiObject } from './api-object';
import { Construct } from '@aws-cdk/core';
import { HorizontalpodautoscalerAutoscalingV2Beta1 } from './.generated/horizontalpodautoscaler-autoscaling-v2beta1';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';

export interface HorizontalPodAutoscalerObjectOptions {
  readonly spec: HorizontalpodautoscalerAutoscalingV2Beta1;
  readonly metadata?: ObjectmetaMetaV1;
}

export class HorizontalPodAutoscalerObject extends ApiObject {
  constructor(scope: Construct, ns: string, options: HorizontalPodAutoscalerObjectOptions) {
    super(scope, ns, {
      apiVersion: 'apps/v1',
      kind: 'HorizontalPodAutoscaler',
      spec: options.spec,
      metadata: options.metadata
    });
  }
}
