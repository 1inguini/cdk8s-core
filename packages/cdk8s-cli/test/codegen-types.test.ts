import { TypeGenerator } from "../lib/codegen-types";
import { promises as fs } from 'fs';
import * as path from 'path';
import { CodeMaker } from "codemaker";
import { jsiiCompile, createWorkdir } from "./util";
import { JSONSchema4 } from "json-schema";

jest.setTimeout(60_000); // 1min

describe('unions', () => {

  which('include primitives', {
    oneOf: [
      { type: "string" },
      { type: "number" }
    ]
  });

});


describe('documentation', () => {

  which('does not render if not defined', {
    type: 'boolean'
  });

  which('renders based on description', {
    description: 'hello, description',
    type: 'string'
  });
  
  which('"*/" is is escaped', {
    description: 'hello */world',
    type: 'string'
  });
  
});

function which(name: string, schema: JSONSchema4) {
  test(name, async () => {
    const gen = new TypeGenerator();
    gen.addType('TestType', schema);
    expect(await generate(gen)).toMatchSnapshot();
  });
}

async function generate(gen: TypeGenerator) {
  const code = new CodeMaker();

  const workdir = await createWorkdir();
  const entrypoint = 'index.ts';

  code.openFile(entrypoint);
  gen.generate(code);
  code.closeFile(entrypoint)
  await code.save(workdir);

  const source = await fs.readFile(path.join(workdir, entrypoint), 'utf-8');

  try {
    await jsiiCompile(workdir);
  } catch (e) {
    console.error(source);
    throw e;
  }

  return source;
}
