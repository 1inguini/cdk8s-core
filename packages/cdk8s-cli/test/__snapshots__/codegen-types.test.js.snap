// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentation "*/" is is escaped 1`] = `
"/**
 * hello _/world
 */
export type TestType = string;

"
`;

exports[`documentation does not render if not defined 1`] = `
"export type TestType = boolean;

"
`;

exports[`documentation renders based on description 1`] = `
"/**
 * hello, description
 */
export type TestType = string;

"
`;

exports[`oneOf 1`] = `
"export class TestType {
    public static fromString(value: string): TestType {
        return new TestType(value);
    }
    public static fromNumber(value: number): TestType {
        return new TestType(value);
    }
    private constructor(value: any) {
        Object.defineProperty(this, 'resolve', { value });
    }
}

"
`;

exports[`unions include primitives 1`] = `
"export class TestType {
    public static fromString(value: string): TestType {
        return new TestType(value);
    }
    public static fromNumber(value: number): TestType {
        return new TestType(value);
    }
    private constructor(value: any) {
        Object.defineProperty(this, 'resolve', { value });
    }
}

"
`;
